userspace/libsinsp/container_engine/docker_common.cpp

```diff
@@ -41,20 +161,106 @@ void docker::parse_json_mounts(const Json::Value &mnt_obj, vector<sinsp_containe
 	}
 }
 
-void docker::set_query_image_info(bool query_image_info)
+void docker_async_source::set_query_image_info(bool query_image_info)
 {
 	m_query_image_info = query_image_info;
 }
 
-bool docker::parse_docker(sinsp_container_manager* manager, sinsp_container_info *container, sinsp_threadinfo* tinfo)
+bool docker::resolve(sinsp_container_manager* manager, sinsp_threadinfo* tinfo, bool query_os_for_missing_info)
+{
+	std::string container_id, container_name;
+	sinsp_container_info *existing_container_info;
+
+	if(!g_docker_info_source)
+	{
+		uint64_t max_wait_ms = 10000;
+		docker_async_source *src = new docker_async_source(docker_async_source::NO_WAIT_LOOKUP, max_wait_ms, manager->get_inspector());
+		g_docker_info_source.reset(src);
+	}
+
+	if(!detect_docker(tinfo, container_id, container_name))
+	{
+		return false;
+	}
+
+	tinfo->m_container_id = container_id;
+
+	existing_container_info = manager->get_container(container_id);
+
+	if(!existing_container_info)
+	{
+		// Add a minimal container_info object where only the
+		// container id, (possibly) name, and a container
+		// image=incomplete is filled in. This may be
+		// overidden later once parse_docker_async completes.
+		sinsp_container_info container_info;
+
+		container_info.m_type = CT_UNKNOWN;
+		container_info.m_id = container_id;
+		container_info.m_name = container_name;
+		container_info.m_image="incomplete";
+		container_info.m_metadata_complete = false;
+
+		manager->add_container(container_info, tinfo);
+
+		existing_container_info = manager->get_container(container_id);
+	}
+
+#ifdef HAS_CAPTURE
+	// Possibly start a lookup for this container info
+	if(!existing_container_info->m_metadata_complete &&
```

Comment:
Can we use the same logic for CRI (testing for `m_metadata_complete`, instead of matching against CT_UNKNOWN)?

Reply:
Yeah I fixed that when I removed CT_UNKNOWN.